### ISSUE TITLE
Unexpected cast to FabricTransportServiceRemotingClient instead of TCommunicationClient

### ISSUE THREAD

**Issue by tophallen** [Azure/service-fabric-issues#454]
I'm trying to override and extend the `FabricTransportServiceRemotingClient` to support a few custom headers, and in the Microsoft.ServiceFabric.Services.Remoting assembly file version 5.7.198.9494 it tries to perform a cast to `FabricTransportServiceRemotingClient` when I believe it should just be `TCommunicationClient` since that is what the `FabricTransportServiceRemotingClientImpl` class needs on that method.

Here is the offending code in `FabricTransportServiceRemotingClientFactory`:
```
Task<OperationRetryControl> ICommunicationClientFactory<IServiceRemotingClient>.
    ReportOperationExceptionAsync(
        IServiceRemotingClient client, 
        ExceptionInformation exceptionInformation, 
        OperationRetrySettings retrySettings, 
        CancellationToken cancellationToken)
{
    return this.impl.ReportOperationExceptionAsync(
        (FabricTransportServiceRemotingClient)client, 
        exceptionInformation, 
        retrySettings, 
        cancellationToken);
}
```

Is this a bug or am I missing something?

**Comment by alexandery** [comment]
I've mentioned this bug to SF guys back in May. I guess it wasn't a priority to fix or something else got in a way.

I've got a workaround for this, which is not pretty - I cast client to my own implementation of IServiceRemotingClient and pass in a cached instance of IServiceRemotingClient:

```c#
ICommunicationClientFactory<IServiceRemotingClient> innerClientFactory

...

return innerClientFactory.ReportOperationExceptionAsync(
                (client as MyFabricTransportServiceRemotingClient).InnerClient,
                exceptionInformation,
                retrySettings,
                cancellationToken);
```

Hope this helps.

**Comment by tophallen** [comment]
Thanks @alexandery that's what I ended up doing as well - seems like it should be an easy fix for the SF team - but for now I pull my inner client using a cast like you said. Being as the `FabricTransportServiceRemotingClientImpl` uses generics I feel like this was probably just a mistake that slipped through, but alas. By the time the code gets there though - I don't need the wrapper I built anyways - it's just for error handling which the user session that I was passing isn't that important there - and if I need I could just put a `TelemetryClient` in that class to handle tracking that information when tracking the error. But you are right; not very pretty.
